packages/medusa/src/services/event-bus.ts

```diff
@@ -265,6 +198,111 @@ export default class EventBusService {
     )
   }
 
+  /**
+   * Adds a function to a list of event subscribers.
+   * @param event - the event that the subscriber will listen for.
+   * @param subscriber - the function to be called when a certain event
+   * happens. Subscribers must return a Promise.
+   * @return this
+   */
+  subscribe(event: string | symbol, subscriber: Subscriber): this {
+    if (typeof subscriber !== "function") {
+      throw new Error("Subscriber must be a function")
+    }
+
+    const observers = this.observers_.get(event) ?? []
+    this.observers_.set(event, [...observers, subscriber])
+
+    return this
+  }
+
+  /**
+   * Adds a function to a list of event subscribers.
+   * @param event - the event that the subscriber will listen for.
+   * @param subscriber - the function to be called when a certain event
+   * happens. Subscribers must return a Promise.
+   * @return this
+   */
+  unsubscribe(event: string | symbol, subscriber: Subscriber): this {
+    if (typeof subscriber !== "function") {
+      throw new Error("Subscriber must be a function")
+    }
+
+    if (this.observers_.get(event)?.length) {
+      const index = this.observers_.get(event)?.indexOf(subscriber)
+      if (index !== -1) {
+        this.observers_.get(event)?.splice(index as number, 1)
+      }
+    }
+
+    return this
+  }
+
+  /**
+   * Calls all subscribers when an event occurs.
+   * @param {string} eventName - the name of the event to be process.
+   * @param data - the data to send to the subscriber.
+   * @param options - options to add the job with
+   * @return the job from our queue
+   */
+  async emit<T>(
+    eventName: string,
+    data: T,
+    options: EmitOptions & { uniqueCacheKey?: string } = {}
+  ): Promise<StagedJob | void> {
```

Comment:
suggestion: Looks like the emit does not return anything anymore. Should we change the signature?